pkg/controller/csr/csr_controller.go

```diff
@@ -0,0 +1,321 @@
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		reqLogger.Info("Signed CSR with name : %v.", csr.Name)
+	}
+
+	return reconcile.Result{}, nil
+}
+
+type systemError struct {
+	message string
+}
+
+func (v systemError) Error() string {
+	return v.message
+}
+
+func (r ReconcileCSR) validate(ctx context.Context, csr *certificatesv1.CertificateSigningRequest) (*x509.Certificate, error) {
+	firstBlock, restBlocks := pem.Decode(csr.Spec.Request)
+	if firstBlock == nil {
+		return nil, fmt.Errorf("cannot parse certificate request for CSR with name %s", csr.Name)
+	}
+	if len(restBlocks) != 0 {
+		return nil, fmt.Errorf("unexpected (multiple) pem blocks for CSR with name %s", csr.Name)
+	}
+
+	certificateRequest, err := x509.ParseCertificateRequest(firstBlock.Bytes)
+	if err != nil {
+		return nil, err
+	}
+	if err = certificateRequest.CheckSignature(); err != nil {
+		return nil, err
+	}
+
+	// The CSRName is formatted as "<secretName>:<podName>".
+	secretName := strings.Split(csr.Name, ":")[0]
+
+	// Validate whether this is a CSR we monitor at all.
+	asset, ok := r.allowedTLSAssets[secretName]
+	if !ok {
+		return nil, fmt.Errorf("this controller is not configured to sign secretName: %s", secretName)
+	}
+
+	// Validate whether the requestor of the CSR is registered for the given CSR
+	if fmt.Sprintf("system:serviceaccount:%s:%s", asset.serviceaccountNamespace, asset.serviceaccountName) != csr.Spec.Username {
+		return nil, fmt.Errorf("invalid requestor %s for CSR with name %s", csr.Spec.Username, csr.Name)
+	}
+
+	// Validate whether the DNS names are permitted for the request.
+	for _, name := range append(certificateRequest.DNSNames, certificateRequest.Subject.CommonName) {
+		var found bool
+		for _, valid := range asset.validDNSNames {
+			if valid == name {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return nil, fmt.Errorf("invalid dns name \"%s\" found in CSR with name %s", name, csr.Name)
+		}
+	}
+
+	if len(certificateRequest.IPAddresses) == 1 {
+		// Validate the IP address that is requested in the CSR matches the pod IP of the requestor.
+		label, ok := csr.Labels[render.AppLabelName]
+		if !ok {
+			return nil, fmt.Errorf("unable to find expected k8s-app label for CSR with name %s", csr.Name)
+		}
+		podList := &corev1.PodList{}
+		err = r.client.List(ctx, podList, &client.ListOptions{
+			LabelSelector: labels.SelectorFromSet(map[string]string{
+				render.AppLabelName: label,
+			}),
+			Namespace: asset.serviceaccountNamespace})
+		if err != nil {
+			return nil, systemError{fmt.Sprintf("unexpected server error while listing pods by label for CSR %s", csr.Name)}
+		}
+		var foundIp bool
+		for _, pod := range podList.Items {
+			if pod.Status.PodIP == certificateRequest.IPAddresses[0].String() {
+				foundIp = true
+				break
+			}
+		}
+		if !foundIp {
+			return nil, fmt.Errorf("invalid pod IP for CSR with name %s", csr.Name)
+		}
+	} else if len(certificateRequest.IPAddresses) > 1 {
```

Comment:
If there are no IP addresses, we issue a cert without IP SAN on them. This is a valid use-case, but with the way we configure our init container it will currently never happen.